### PR TITLE
Fix TLS version negotiation based on sigalg availability

### LINKED ISSUE #25697: TLS1.3 with brainpool version incompatibiliity
When an openssl-3.0 client wants to support brainpool server certificates,
but also allow TLS1.3 with RSA certificates, everything works fine, when
both client and server enable brainpool groups, in addition to nist groups.
So the client has to use this command line:
` ./openssl s_client -groups P-256:brainpoolP256r1`
and the openssl-3.0 server uses this command line for brainpool:
`./openssl s_server -cert ../test/certs/server-ecdsa-brainpoolP256r1-cert.pem -key ../test/certs/server-ecdsa-brainpoolP256r1-key.pem -groups P-256:brainpoolP256r1`
or this command line for RSA:
`./openssl s_server`
everything works fine, RSA use TLS1.3 and ECC use TLS1.2.
Also if both server an client use openssl-3.2 everything works as well,
except the ECC use TLS1.3, as expected.
BUT the openssl-3.0 client cannot connect to the openssl-3.2 server
when the server uses brainpool:
```
$ ./openssl s_client -groups P-256:brainpoolP256r1 -trace
CONNECTED(00000003)
Sent Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Handshake (22)
  Length = 327
    ClientHello, Length=323
      client_version=0x303 (TLS 1.2)
      Random:
        gmt_unix_time=0xD3B13943
        random_bytes (len=28): 24530A97534115EA57B17C77ABAC33F909733FDBAA73C61AD594C525
      session_id (len=32): 397A7086775ACCDB82A75E2B19B7FE6459BF3088B322A7A5D8EAC17791A76D40
      cipher_suites (len=62)
        {0x13, 0x02} TLS_AES_256_GCM_SHA384
        {0x13, 0x03} TLS_CHACHA20_POLY1305_SHA256
        {0x13, 0x01} TLS_AES_128_GCM_SHA256
        {0xC0, 0x2C} TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
        {0xC0, 0x30} TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
        {0x00, 0x9F} TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
        {0xCC, 0xA9} TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
        {0xCC, 0xA8} TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
        {0xCC, 0xAA} TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256
        {0xC0, 0x2B} TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
        {0xC0, 0x2F} TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
        {0x00, 0x9E} TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
        {0xC0, 0x24} TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
        {0xC0, 0x28} TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
        {0x00, 0x6B} TLS_DHE_RSA_WITH_AES_256_CBC_SHA256
        {0xC0, 0x23} TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
        {0xC0, 0x27} TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
        {0x00, 0x67} TLS_DHE_RSA_WITH_AES_128_CBC_SHA256
        {0xC0, 0x0A} TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
        {0xC0, 0x14} TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
        {0x00, 0x39} TLS_DHE_RSA_WITH_AES_256_CBC_SHA
        {0xC0, 0x09} TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
        {0xC0, 0x13} TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0x33} TLS_DHE_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0x9D} TLS_RSA_WITH_AES_256_GCM_SHA384
        {0x00, 0x9C} TLS_RSA_WITH_AES_128_GCM_SHA256
        {0x00, 0x3D} TLS_RSA_WITH_AES_256_CBC_SHA256
        {0x00, 0x3C} TLS_RSA_WITH_AES_128_CBC_SHA256
        {0x00, 0x35} TLS_RSA_WITH_AES_256_CBC_SHA
        {0x00, 0x2F} TLS_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0xFF} TLS_EMPTY_RENEGOTIATION_INFO_SCSV
      compression_methods (len=1)
        No Compression (0x00)
      extensions, length = 188
        extension_type=server_name(0), length=14
          0000 - 00 0c 00 00 09 6c 6f 63-61 6c 68 6f 73 74      .....localhost
        extension_type=ec_point_formats(11), length=4
          uncompressed (0)
          ansiX962_compressed_prime (1)
          ansiX962_compressed_char2 (2)
        extension_type=supported_groups(10), length=6
          secp256r1 (P-256) (23)
          brainpoolP256r1 (26)
        extension_type=session_ticket(35), length=0
        extension_type=encrypt_then_mac(22), length=0
        extension_type=extended_master_secret(23), length=0
        extension_type=signature_algorithms(13), length=42
          ecdsa_secp256r1_sha256 (0x0403)
          ecdsa_secp384r1_sha384 (0x0503)
          ecdsa_secp521r1_sha512 (0x0603)
          ed25519 (0x0807)
          ed448 (0x0808)
          rsa_pss_pss_sha256 (0x0809)
          rsa_pss_pss_sha384 (0x080a)
          rsa_pss_pss_sha512 (0x080b)
          rsa_pss_rsae_sha256 (0x0804)
          rsa_pss_rsae_sha384 (0x0805)
          rsa_pss_rsae_sha512 (0x0806)
          rsa_pkcs1_sha256 (0x0401)
          rsa_pkcs1_sha384 (0x0501)
          rsa_pkcs1_sha512 (0x0601)
          ecdsa_sha224 (0x0303)
          rsa_pkcs1_sha224 (0x0301)
          dsa_sha224 (0x0302)
          dsa_sha256 (0x0402)
          dsa_sha384 (0x0502)
          dsa_sha512 (0x0602)
        extension_type=supported_versions(43), length=9
          TLS 1.3 (772)
          TLS 1.2 (771)
          TLS 1.1 (770)
          TLS 1.0 (769)
        extension_type=psk_key_exchange_modes(45), length=2
          psk_dhe_ke (1)
        extension_type=key_share(51), length=71
            NamedGroup: secp256r1 (P-256) (23)
            key_exchange:  (len=65): 04B9F00C469BB69090E6B7E8FC0EF518A600EEA9950825CFAFBC88DCF2B0EE218100ABD95B8C6CFAA29A7302332E42FCB07286C0653C823E5896981ED97C363D0D

Received Record
Header:
  Version = TLS 1.2 (0x303)
  Content Type = Alert (21)
  Length = 2
    Level=fatal(2), description=handshake failure(40)

40778BA11E7F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1605:SSL alert number 40
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 332 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```
Reason is the openssl-3.2 server chooses TLS1.3 although there is no
signature algorithm `ecdsa_brainpoolP256r1_sha256` advertised by the
openssl-3.0 client, so that is bound to fail.
As a workaround either the openssl-3.0 client or the openssl-3.2 server could
disable tls1_3 but that is not desirable for the RSA server.

### PR DESCRIPTION
**Pull request by bernd-edlinger**
If a server is using a brainpool certificate and if we can negotiate either TLS1.2 or TLS1.3, but the necessary sigalg is missing, TLS1.3 will fail immediately, while TLS1.2 may work. So use the signature_algorithms extension of the ClientHello as a hint which protocol to choose.

Fixes #25697

### CODE REVIEW

**Comment by davidben** [comment]
Hmm. I haven't looked carefully at how you all implement this, but I suspect that either this does not actually work, or that OpenSSLs where `is_tls13_capable` checks are reachable are silently less secure than those without. I.e. this will make brainpool TLS 1.3 less secure than standard TLS 1.3.

TLS 1.3 added this downgrade protection signal to the ServerHello random. You're supposed to send it whenever the server is TLS-1.3-capable but picked TLS 1.2 for whatever reason. That will then cause a TLS-1.3-capable client to reject the connection.
https://www.rfc-editor.org/rfc/rfc8446#section-4.1.3

Now, this signal is a workaround for TLS 1.2 ServerKeyExchange failing to sign the entire transcript. The signature payload doesn't contain any context outside the two randoms. An attacker is free to arbitrarily modify the ClientHello and see what ServerKeyExchange signature they can get out. That means if _any_ ClientHello can induce a server to behave as if it doesn't support TLS 1.3, the attacker can do that and bypass this downgrade protection check.

This is relevant if, for example, the client supports [TLS False Start](https://www.rfc-editor.org/rfc/rfc7918.html) over TLS 1.2, because then it means the Finished check doesn't work.

(This is a bit unfortunate because it means that this TLS 1.3 hardening measure is kind of a hair trigger. It does not tolerate the existence of _any_ non-version-related reason to pick TLS 1.2 over TLS 1.3. In hindsight, the WG probably should have made a `sign_whole_transcript` TLS 1.2 extension that could be negotiated more unconditionally than TLS 1.3, and then used the random hack for it instead. I suppose it's not too late to do that, though I dunno how much appetite there is to add it to TLS 1.2 now.)

**Comment by bernd-edlinger** [comment]
Thanks for the background, this is meant to to illustrate a possible alternative to #25733 
And Yes, this kind of relies on the Finished verify_data to authenticate the handshake.
The situation with brainpool is this:
Due to my #7485 which added the ability to use TLS1.3 with brainpool it is now possible for server and client
to use these certificates, by default, when both have the latest openssl version, that is 3.2.1 or newer, which
was released in january of this year.  That is of course nice.
BUT it is not interoperable with applications that are written for previous version, where only TLS1.2 was able
to use brainpool, when both client and server explicitly add brainpool to the groups.
The client has no problem, if can now support both TLS1.3 and TLS1.2, while it previously only supported TLS1.2,
that is good.
But if the server tries the support both TLS1.3 and TLS1.2 by adding brainpool to the groups, this does not work.
It will always negotiate TLS1.3, except with super old clients, prior to 1.1.1 where only TLS1.2 was supported.

The only way out for my application, is now to roll out an update, that removes the brainpool signatures
from the server's sigalgs, but keeps the brainpool signatures for the client's sigalgs.  That must be done
in a way that does runtime detect the openssl version, so use the return code of the `SSL_CTX_set1_sigalgs_list`
see if the :ecdsa_brainpoolP256r1_sha256 is understood so we have 3.2.1 or newer runtime, and if not
undo the damage by setting sigalgs list without brainpool.  The server on the other hand must explicitly
use a sigalgs list without brainpool signatures, and might try to use a  -client_sigalgs which includes brainpool.
I need to roll out that before O/S vendors start to roll out the openssl 3.2, and keep that for 4-5 years,
until openssl-3.2 is available everywhere.

So I am wondering if it is better to permanently disable TLS1.3 or to accept this theoretic downgrade issue
for brainpool curves.
In other words, do the signatures from the TLS1.2 ServerKeyExchange expose any threat to the TLS1.3 protocol?
